### PR TITLE
ci: Physmon for event simulation

### PR DESCRIPTION
**Pull request by andiwand**
All our physics monitors depend on the simulation which is not directly monitored at the moment. In this PR I add a first monitoring version for Fatras and Geant4.

The monitoring is not optimal yet but should be sufficient if something changes or breaks for simulation.

Possible improvements:
- record deltas for momenta / eta / phi and so on instead of initial and final

Blocked by
- https://github.com/acts-project/acts/pull/2241
- https://github.com/acts-project/acts/pull/2260
- https://github.com/acts-project/acts/pull/2262
- https://github.com/acts-project/acts/pull/2263
- https://github.com/acts-project/acts/pull/2264
- https://github.com/acts-project/acts/pull/2265
- https://github.com/acts-project/acts/pull/2266

### CODE REVIEW

File: Examples/Algorithms/Geant4/src/ParticleKillAction.cpp [hunk 1]
```diff
@@ -34,16 +34,20 @@ void ActsExamples::ParticleKillAction::UserSteppingAction(const G4Step* step) {
 
   const auto pos = convertLength * track->GetPosition();
   const auto time = convertTime * track->GetGlobalTime();
+  const bool secondary =
+      track->GetDynamicParticle()->GetPrimaryParticle() != nullptr;
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Algorithms/Geant4/src/ParticleKillAction.cpp]: nice catch!

File: Core/src/Geometry/TrackingGeometry.cpp [hunk 2]
```diff
@@ -53,7 +53,12 @@ const Acts::TrackingVolume* Acts::TrackingGeometry::lowestTrackingVolume(
 
 const Acts::TrackingVolume* Acts::TrackingGeometry::highestTrackingVolume()
     const {
-  return (m_world.get());
+  return m_world.get();
+}
+
+std::shared_ptr<const Acts::TrackingVolume>
```
Comment (paulgessinger) [inline Core/src/Geometry/TrackingGeometry.cpp:59]: Probably unnecessary but

```suggestion
const std::shared_ptr<const Acts::TrackingVolume>&
```

File: Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp [hunk 3]
```diff
@@ -102,15 +102,15 @@ void ActsExamples::ParticleTrackingAction::PostUserTrackingAction(
...
+  constexpr double convertTime = Acts::UnitConstants::ns / CLHEP::ns;
   constexpr double convertLength = Acts::UnitConstants::mm / CLHEP::mm;
   constexpr double convertEnergy = Acts::UnitConstants::GeV / CLHEP::GeV;
 
   // Get all the information from the Track
   const G4ParticleDefinition* particleDef = aTrack.GetParticleDefinition();
   G4int pdg = particleDef->GetPDGEncoding();
   G4double charge = particleDef->GetPDGCharge();
-  G4double mass = particleDef->GetPDGMass();
+  G4double mass = convertEnergy * particleDef->GetPDGMass();
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:113]: Was this the unit bug you mentioned?

File: Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp [hunk 4]
```diff
@@ -76,6 +76,7 @@ void ActsExamples::SimParticleTranslation::GeneratePrimaries(G4Event* anEvent) {
   std::optional<Acts::Vector4> lastVertex;
 
   constexpr double convertLength = CLHEP::mm / Acts::UnitConstants::mm;
+  constexpr double convertTime = CLHEP::ns / Acts::UnitConstants::ns;
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp:79]: Should we centralize these definitions at some point?

File: Examples/Io/Root/include/ActsExamples/Io/Root/RootParticleWriter.hpp [hunk 5]
```diff
@@ -82,17 +82,17 @@ class RootParticleWriter final : public WriterT<SimParticleContainer> {
   /// Production process type, i.e. what generated the particle.
   std::vector<uint32_t> m_process;
   /// Production position components in mm.
-  std::vector<float> m_vx;
-  std::vector<float> m_vy;
-  std::vector<float> m_vz;
```
Comment (paulgessinger) [inline Examples/Io/Root/include/ActsExamples/Io/Root/RootParticleWriter.hpp:87]: Any particular motivation for this?

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/ParticleKillAction.hpp [hunk 6]
```diff
@@ -25,15 +25,16 @@ namespace ActsExamples {
...
 /// the simulation process.
-///
-/// It checks whether the particle can be killed according when its position
-/// exceeds the configured values for |z| or r.
 class ParticleKillAction : public G4UserSteppingAction {
  public:
   /// Configuration of the Stepping action
   struct Config {
+    /// particles outside this volume will be terminated
     std::shared_ptr<const Acts::Volume> volume;
```
Comment (paulgessinger) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/ParticleKillAction.hpp:33]: Is the possibility to have ownership of the volume here just to have the ability to define a custom volume?

File: Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp (hunk 3 above)
Comment (andiwand) [reply Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:113]: exactly I should make this clear in the description

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/ParticleKillAction.hpp (hunk 6 above)
Comment (andiwand) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/ParticleKillAction.hpp:33]: yes initially the idea was to pass an arbitrary volume here. by default we pass the detector volume now

File: Examples/Io/Root/include/ActsExamples/Io/Root/RootParticleWriter.hpp (hunk 5 above)
Comment (andiwand) [reply Examples/Io/Root/include/ActsExamples/Io/Root/RootParticleWriter.hpp]: this was coming from https://github.com/acts-project/acts/pull/2241 I can revert that now

File: Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp (hunk 4 above)
Comment (andiwand) [reply Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp:79]: that would be good I think - but I would rather do that in a separate PR

**Review by paulgessinger — APPROVED** [review]
LGTM!

**Comment by andiwand** [comment]
hitting some new FPEs

```
15:55:42    Sequencer      INFO      -----------------------------------
15:55:42    Sequencer      INFO      FPE summary for Writer: RootTrajectorySummaryWriter
15:55:42    Sequencer      INFO      -----------------------------------
15:55:42    Sequencer      INFO      - FLTINV: (1 times) 
 0# ActsExamples::RootTrajectorySummaryWriter::writeT(ActsExamples::AlgorithmContext const&, std::vector<ActsExamples::Trajectories,
std::allocator<ActsExamples::Trajectories> > const&) at /builds/acts/ci-bridge/src/Examples/Io/Root/src/RootTrajectorySummaryWriter.cpp:350
 1# ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const::{lambda(tbb::detail::d1::blocked_range<unsigned long>
const&)#1}::operator()(tbb::detail::d1::blocked_range<unsigned long> const&) const at /builds/acts/ci-
bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:505
 2# ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const at /builds/acts/ci-bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:466
 3# ActsExamples::Sequencer::run() at /builds/acts/ci-bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:565
 4# pybind11::cpp_function::initialize<pybind11_init_ActsPythonBindings(pybind11::module_&)::{lambda(ActsExamples::Sequencer&)#4}, void,
ActsExamples::Sequencer&, pybind11::name, pybind11::is_method,
pybind11::sibling>(pybind11_init_ActsPythonBindings(pybind11::module_&)::{lambda(ActsExamples::Sequencer&)#4}&&, void (*)(ActsExamples::Sequencer&),
pybind11::name const&, pybind11::is_method const&, pybind11::sibling
const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call&) at /builds/acts/ci-
bridge/build/_deps/pybind11-src/include/pybind11/pybind11.h:224
 5# pybind11::cpp_function::dispatcher(_object*, _object*, _object*) at /builds/acts/ci-
bridge/build/_deps/pybind11-src/include/pybind11/pybind11.h:929
 6# 0x000055FCBFDA0B5E in /usr/bin/python3
 7# _PyObject_MakeTpCall in /usr/bin/python3

15:55:42    Sequencer      INFO      -----------------------------------
15:55:42    Sequencer      INFO      FPE summary for Writer: RootTrajectorySummaryWriter
15:55:42    Sequencer      INFO      -----------------------------------
15:55:42    Sequencer      INFO      - FLTINV: (1 times) 
 0# ActsExamples::RootTrajectorySummaryWriter::writeT(ActsExamples::AlgorithmContext const&, std::vector<ActsExamples::Trajectories,
std::allocator<ActsExamples::Trajectories> > const&) at /builds/acts/ci-bridge/src/Examples/Io/Root/src/RootTrajectorySummaryWriter.cpp:350
 1# ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const::{lambda(tbb::detail::d1::blocked_range<unsigned long>
const&)#1}::operator()(tbb::detail::d1::blocked_range<unsigned long> const&) const at /builds/acts/ci-
bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:505
 2# ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const at /builds/acts/ci-bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:466
 3# ActsExamples::Sequencer::run() at /builds/acts/ci-bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:565
 4# pybind11::cpp_function::initialize<pybind11_init_ActsPythonBindings(pybind11::module_&)::{lambda(ActsExamples::Sequencer&)#4}, void,
ActsExamples::Sequencer&, pybind11::name, pybind11::is_method,
pybind11::sibling>(pybind11_init_ActsPythonBindings(pybind11::module_&)::{lambda(ActsExamples::Sequencer&)#4}&&, void (*)(ActsExamples::Sequencer&),
pybind11::name const&, pybind11::is_method const&, pybind11::sibling
const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call&) at /builds/acts/ci-
bridge/build/_deps/pybind11-src/include/pybind11/pybind11.h:224
 5# pybind11::cpp_function::dispatcher(_object*, _object*, _object*) at /builds/acts/ci-
bridge/build/_deps/pybind11-src/include/pybind11/pybind11.h:929
 6# 0x000055FCBFDA0B5E in /usr/bin/python3
 7# _PyObject_MakeTpCall in /usr/bin/python3

15:55:42    Sequencer      ERROR     Encountered 2 unmasked FPEs
```

need to look into that

**Review by paulgessinger — APPROVED** [review]
Looks goo as far as I can tell.

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838697) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838696) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838695) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838694) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsTrkSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838693) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838692) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838691) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838690) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838689) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838688) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838687) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838686) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838685) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838684) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838683) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838682) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/30838681) |

**Pull request merged**